lib/database/users.ts

```diff
@@ -0,0 +1,115 @@
+import { client } from "../db";
+import { updateUserSchema, User, userSchema, UpdateUser } from "../../models/users";
+import { array, InferType } from "yup";
+
+const userArraySchema = array(userSchema).defined();
+type userArrayType = InferType<typeof userArraySchema>;
+
+// create a user in the database with given parameters.
+const createUser = async (
+  id: string,
+  first_name: string,
+  last_name: string,
+  email: string,
+  role: string,
+  address: string,
+  phone_number?: string
+): Promise<User> => {
+  const query = {
+    text:
+      "INSERT INTO users(id, first_name, last_name, email, role, address, phone_number) VALUES($1, $2, $3, $4, $5, $6, $7)" +
+      "RETURNING  id, first_name, last_name, email, role, phone_number, address",
+    values: [id, first_name, last_name, email, role, address, phone_number],
+  };
+
+  const res = await client.query(query);
+  // basic error checking
+  if (res.rowCount != 1) throw Error("Database returned the incorrect number of rows");
+
+  let user: User;
+
+  try {
+    user = await userSchema.validate(res.rows[0]);
+  } catch {
+    throw Error("Error on return from database");
+  }
+
+  return user;
+};
+
+// updates user given new parameters
+const updateUser = async (
+  id: string,
+  first_name?: string,
+  last_name?: string,
+  email?: string,
+  role?: string,
+  address?: string,
+  phone_number?: string,
+): Promise<User | null> => {
+  const query = {
+      text:"UPDATE users " +
+      "SET first_name = COALESCE($2, first_name), "+ 
+      "last_name = COALESCE($3, last_name), " +
+      "email = COALESCE($4, email), " +
+      "role = COALESCE($5, role), " +
+      "address = COALESCE($6, address), " +
+      "phone_number = COALESCE($7, phone_number) " + 
+      "WHERE id=$1",
+      values: [id, first_name, last_name, email, role, address, phone_number],
+  }
+        
+    try {
+        const res = await client.query(query);
+    }
+    catch{
+        throw Error("Internal Server Error");
```

Comment:
since we probably will be logging things, lets make this error message "Error on update user"

Reply:
should I also change the status codes in the handler for this error?

Reply:
Nah. It's still an internal server error but the Error should be a bit more specific so when we log the errors in the database we know where the error came from